pkg/agent/ipassigner/responder/arp_responder.go

```diff
@@ -65,40 +57,88 @@ func (r *arpResponder) RemoveIP(ip net.IP) error {
 		return fmt.Errorf("only IPv4 is supported")
 	}
 	if r.deleteIP(ip) {
-		klog.InfoS("Removed IP from ARP responder", "ip", ip, "interface", r.iface.Name)
+		klog.InfoS("Removed IP from ARP responder", "ip", ip, "interface", r.linkName)
 	}
 	return nil
 }
 
-func (r *arpResponder) handleARPRequest() error {
-	pkt, _, err := r.conn.Read()
+func (r *arpResponder) handleARPRequest(client *arp.Client, iface *net.Interface) error {
+	pkt, _, err := client.Read()
 	if err != nil {
 		return err
 	}
 	if pkt.Operation != arp.OperationRequest {
 		return nil
 	}
 	if !r.isIPAssigned(pkt.TargetIP) {
-		klog.V(4).InfoS("Ignored ARP request", "ip", pkt.TargetIP, "interface", r.iface.Name)
+		klog.V(4).InfoS("Ignored ARP request", "ip", pkt.TargetIP, "interface", r.linkName)
 		return nil
 	}
-	if err := r.conn.Reply(pkt, r.iface.HardwareAddr, pkt.TargetIP); err != nil {
+	if err := client.Reply(pkt, iface.HardwareAddr, pkt.TargetIP); err != nil {
 		return fmt.Errorf("failed to reply ARP packet for IP %s: %v", pkt.TargetIP, err)
 	}
-	klog.V(4).InfoS("Sent ARP response", "ip", pkt.TargetIP, "interface", r.iface.Name)
+	klog.V(4).InfoS("Sent ARP response", "ip", pkt.TargetIP, "interface", r.linkName)
 	return nil
 }
 
 func (r *arpResponder) Run(stopCh <-chan struct{}) {
+	// The responder instance is created by the factory and can be shared by multiple callers.
+	// Using once.Do here ensures it is started only once.
+	r.once.Do(func() {
+		wait.NonSlidingUntil(func() {
+			r.dialAndHandleRequests(stopCh)
+		}, time.Second, stopCh)
+	})
+	<-stopCh
+}
+
+func (r *arpResponder) dialAndHandleRequests(stopCh <-chan struct{}) {
+	transportInterface, err := net.InterfaceByName(r.linkName)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get interface by name", "deviceName", r.linkName)
+		return
+	}
+	client, err := arp.Dial(transportInterface)
+	if err != nil {
+		klog.ErrorS(err, "Failed to dial ARP client", "deviceName", r.linkName)
+		return
+	}
+	reloadCh := make(chan struct{})
+
+	klog.InfoS("ARP responder started", "interface", transportInterface.Name, "index", transportInterface.Index)
+	defer klog.InfoS("ARP responder stopped", "interface", transportInterface.Name, "index", transportInterface.Index)
+
+	go func() {
+		defer client.Close()
+		defer close(reloadCh)
+
+		for {
+			select {
+			case <-stopCh:
+				return
+			case <-r.linkEventCh:
+				newTransportInterface, err := net.InterfaceByName(r.linkName)
+				if err != nil {
+					klog.ErrorS(err, "Failed to get interface by name", "name", r.linkName)
+					continue
+				}
+				if newTransportInterface.Index != transportInterface.Index {
+					klog.InfoS("Transport interface index changed. Restarting ARP responder", "name", transportInterface.Name, "oldIndex", r.linkIndex, "newIndex", r.linkIndex)
+					return
+				}
+				klog.InfoS("Transport interface not changed")
```

Comment:
It is an event we are interested in, doesn't seem worth being logged at default level.